### PR TITLE
Dobrescu HW03

### PR DESCRIPTION
**Pull request by dobreasca**

### CODE REVIEW

File: Main.hs [hunk 1]
```diff
@@ -1,39 +1,106 @@
...
-  (==) = undefined
+  (==) :: Expr -> Expr -> Bool
+  Num a == Num b = a == b
+  Sqrt a == Sqrt b = a == b
+  Minus a b == Minus c d = a == c && b == d
+  Plus a b == Plus c d = a == c && b == d
+  Multi a b == Multi c d = a == c && b == d
+  Div a b == Div c d = a == c && b == d
+  Power a b == Power c d = a == c && b == d
+  _ == _ = False
```
Comment (kajigor) [inline Main.hs:35]: Whenever you have a standard implementation of `Eq`, it's better to derive it

File: Main.hs [hunk 2]
```diff
@@ -1,39 +1,106 @@
...
 instance Show Error where
-  show = undefined
+  show (DivisionByZero a b) = printf "Error: tried to do devision by zero: %s / %s" (show a) (show b)
+  show (SqrtOfNegNumber a) = printf "Error: tried to find the square root of a negative number: sqrt %s" (show a)
 
 instance Eq Error where
-  (==) = undefined
+  DivisionByZero _ _ == DivisionByZero _ _ = True
+  SqrtOfNegNumber _ == SqrtOfNegNumber _ = True
+  _ == _ = False
```
Comment (kajigor) [inline Main.hs:48]: I would argue that `SqrtOfNegNumber (Num (-1))` and `SqrtOfNegNumber (Minus (Num 1) (Num 2))` are different errors

File: Main.hs [hunk 3]
```diff
@@ -1,39 +1,106 @@
...
+eval (Div a b) = do
+  val1 <- eval a
+  val2 <- eval b
+  if val2 == 0
+    then Left (DivisionByZero a b)
+    else Right (val1 / val2)
+eval (Power a b) = do
+  val1 <- eval a
+  val2 <- eval b
+  Right (val1 ** val2)
```
Comment (kajigor) [inline Main.hs:76]: You have a lot of code duplication here. Could you introduce a helper function to get rid of it?

**Review by kajigor — CHANGES_REQUESTED** [review]

File: Main.hs (hunk 3 above)
Comment (dobreasca) [reply Main.hs]: added the helper function and changed the code according the previous comments

File: Main.hs [hunk 4]
```diff
@@ -1,39 +1,97 @@
...
+  | SqrtOfNegNumber Expr
 
 instance Show Error where
-  show = undefined
+  show (DivisionByZero a b) = printf "Error: tried to do division by zero: %s / %s" (show a) (show b)
+  show (SqrtOfNegNumber a) = printf "Error: tried to find the square root of a negative number: sqrt %s" (show a)
 
 instance Eq Error where
-  (==) = undefined
+  DivisionByZero a _ == DivisionByZero b _ = a == b
```
Comment (kajigor) [inline Main.hs:46]: Well... You'd be surprised, but `DivisionByZero (Num 1) (Num 0)` and `DivisionByZero (Num 1) (Minus (Num 1) (Num 1))` are also different.

**Comment by kajigor** [comment]
Helper is ok, but there are still issues with `Eq` instances.

**Pull request closed without merging**